analysis/src/diff.rs

```diff
@@ -0,0 +1,156 @@
+            "https://crates.io/api/v1/crates/{}/{}/download",
+            name, version
+        );
+        let dest_file = format!("{}-{}-cratesio", name, version);
+        Ok(self.download_file(download_path, dest_file)?)
+    }
+
+    fn get_github_repo(&self, name: &String, url: &str) -> Result<Repository> {
+        let dest_file = format!("{}-source", name);
+        let dest_path = self.dir.path().join(&dest_file);
+        let repo = Repository::clone(url, dest_path)?;
+        Ok(repo)
+    }
+
+    fn download_file(&self, download_path: String, dest_file: String) -> Result<PathBuf> {
+        // Destination directory to contain downloded files
+        let dest_path = self.dir.path().join(&dest_file);
+
+        // check if destination directory exists, if not proceed
+        if !dest_path.exists() {
+            // First download the file as tar_gz
+            let targz_path = self.dir.path().join(format!("{}.targ.gz", dest_file));
+            let mut targz_file = File::create(&targz_path)?;
+            let mut response = self.client.get(download_path).send()?;
+            copy(&mut response, &mut targz_file)?;
+
+            // Then decompress the file
+            Self::decompress_targz(&targz_path, &dest_path)?;
+        }
+
+        // Get the only directory within dest_path where files are unpacked
+        let entries: Vec<DirEntry> = read_dir(dest_path)?
+            .filter_map(|entry| entry.ok())
+            .collect();
+        if entries.len() != 1 {
+            return Err(anyhow!("Error in locating directory for unpacked files"));
+        }
+
+        // Return the directory containing unpacked files
+        Ok(entries[0].path())
+    }
+
+    fn decompress_targz(targz_path: &PathBuf, dest_path: &PathBuf) -> Result<()> {
+        let tar_gz = File::open(targz_path)?;
+        let tar = GzDecoder::new(tar_gz);
+        let mut archive = Archive::new(tar);
+        archive.unpack(dest_path)?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    fn get_test_diff_analyzer() -> DiffAnalyzer {
+        DiffAnalyzer::new().unwrap()
+    }
+
+    #[test]
+    #[ignore]
+    fn test_diff_download_file() {
+        let diff_analyzer = get_test_diff_analyzer();
+        let name = "libc";
+        let version = "0.2.97";
+        let path = diff_analyzer
+            .download_file(
+                format!(
+                    "https://crates.io//api/v1/crates/{}/{}/download",
+                    name, version
+                ),
+                format!("{}-{}", name, version),
+            )
+            .unwrap();
+        // println!("{:?}", path);
+        assert_eq!(path.exists(), true);
+    }
+
+    #[test]
+    fn test_diff_crate_source_path() {
+        let diff_analyzer = get_test_diff_analyzer();
+        let name = String::from("libc");
+        let version = String::from("0.2.97");
+        let path = diff_analyzer.get_cratesio_version(&name, &version).unwrap();
```

Comment:
You'll see that change the function signature will stop you needing to make a full `String` here for name and version